From 5d9126345d22b5227e89bbc2f8c10d693331f749 Mon Sep 17 00:00:00 2001
From: Radhey Shyam Pandey <radhey.shyam.pandey@xilinx.com>
Date: Thu, 19 Nov 2015 20:06:23 +0530
Subject: [PATCH 300/793] media: xilinx: Fix TPG moving box and cross-hair
 states

In TPG v7 foreground pattern moving box and
cross-hair are mutually exclusive. i.e
Disable cross-hair if moving box is enabled.

Signed-off-by: Radhey Shyam Pandey <radheys@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/media/platform/xilinx/xilinx-tpg.c |   12 ++++++++++++
 1 files changed, 12 insertions(+), 0 deletions(-)

diff --git a/drivers/media/platform/xilinx/xilinx-tpg.c b/drivers/media/platform/xilinx/xilinx-tpg.c
index a7327fc..5b257e3 100644
--- a/drivers/media/platform/xilinx/xilinx-tpg.c
+++ b/drivers/media/platform/xilinx/xilinx-tpg.c
@@ -140,6 +140,8 @@ struct xtpg_device {
 	struct v4l2_ctrl *hblank;
 	struct v4l2_ctrl *vblank;
 	struct v4l2_ctrl *pattern;
+	struct v4l2_ctrl *moving_box;
+	struct v4l2_ctrl *cross_hair;
 	bool streaming;
 	bool is_hls;
 
@@ -546,10 +548,14 @@ static int xtpg_hls_s_ctrl(struct v4l2_ctrl *ctrl)
 				 XTPG_PATTERN_MASK, ctrl->val);
 		return 0;
 	case V4L2_CID_XILINX_TPG_CROSS_HAIRS:
+		if (ctrl->val)
+			__v4l2_ctrl_s_ctrl(xtpg->moving_box, 0x0);
 		xvip_write(&xtpg->xvip, XTPG_HLS_FG_PATTERN,
 			   ctrl->val << XTPG_HLS_FG_CROSS_HAIR_SHIFT);
 		return 0;
 	case V4L2_CID_XILINX_TPG_MOVING_BOX:
+		if (ctrl->val)
+			__v4l2_ctrl_s_ctrl(xtpg->cross_hair, 0x0);
 		xvip_write(&xtpg->xvip, XTPG_HLS_FG_PATTERN, ctrl->val);
 		return 0;
 	case V4L2_CID_XILINX_TPG_COLOR_MASK:
@@ -1183,6 +1189,12 @@ static int xtpg_probe(struct platform_device *pdev)
 		ret = xtpg->ctrl_handler.error;
 		goto error;
 	}
+
+	xtpg->moving_box = v4l2_ctrl_find(&xtpg->ctrl_handler,
+					  V4L2_CID_XILINX_TPG_MOVING_BOX);
+	xtpg->cross_hair = v4l2_ctrl_find(&xtpg->ctrl_handler,
+					  V4L2_CID_XILINX_TPG_CROSS_HAIRS);
+
 	subdev->ctrl_handler = &xtpg->ctrl_handler;
 
 	xtpg_update_pattern_control(xtpg, true, true);
-- 
1.7.5.4

